warehouse/utils/otp.py

```diff
@@ -0,0 +1,59 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import pyotp
+
+
+def generate_totp_secret():
+    """
+    Generates a secret for time-based OTP.
+
+    The default secret length is 160 bits, as per RFC4226:
+    https://tools.ietf.org/html/rfc4226#section-4
+
+    Given the base32 encoding, it is represented as a 32-character string, i.e.:
+    5 bits per character * 32 characters = 160 bits.
+
+    TODO:
+    consider using os.urandom(); this, however, is likely to cause
+    compatibility issues unless base32 vocabulary is used.
+    """
+    return pyotp.random_base32(length=32)
```

Comment:
Looking at this function, it's not actually being base32 encoded, it's just generating a 16 length random string which just happens to also be valid base32. If this just needs a be a base32 encoded string, then I'd prefer to do something like:

```python
import base64
import os

def generate_totp_secret():
    return base64.b32encode(os.urandom(32)).decode("ascii")
```

This matches how we've done things in ``warehouse/utils/crypto.py``.